src/lib/core/CHIPError.h

```diff
@@ -24,52 +24,158 @@
- *
  */
 
 #pragma once
 
 #include <core/CHIPConfig.h>
 
+#include <inttypes.h>
 #include <stdint.h>
+#include <type_traits>
 
-// clang-format off
+namespace chip {
 
 /**
- *  The basic type for all CHIP errors.
- *
- *  @brief
- *    This is defined to a platform- or system-specific type.
- *
- */
-typedef CHIP_CONFIG_ERROR_TYPE CHIP_ERROR;
+ * This is a helper class for managing `CHIP_ERROR` numbers.
+ *
+ * At the top level, an error belongs to a `Range` and has an integral Value whose meaning depends on the `Range`.
+ * One, `Range::kSDK`, is used for the CHIP SDK's own errors; others encapsulate error codes from external sources
+ * (e.g. libraries, OS) into a `CHIP_ERROR`.
+ *
+ * CHIP SDK errors inside `Range::kSDK` consist of a component identifier given by `SdkPart` and an arbitrary small
+ * integer Code.
+ */
+class ChipError
+{
+public:
+    using BaseType = uint32_t;
+
+    /// `printf` format for error numbers. This is a C macro in order to allow for string literal concatenation.
+#define CHIP_ERROR_FORMAT PRIx32
+
+    /// Top-level error classification.
+    enum class Range : uint8_t
+    {
+        kSDK        = 0x0, ///< CHIP SDK errors.
+        kOS         = 0x1, ///< Encapsulated OS errors, other than POSIX errno.
+        kPOSIX      = 0x2, ///< Encapsulated POSIX errno values.
+        kLwIP       = 0x3, ///< Encapsulated LwIP errors.
+        kOpenThread = 0x4, ///< Encapsulated OpenThread errors.
+        kPlatform   = 0x5, ///< Platform-defined encapsulation.
+    };
+
+    /// Secondary classification of errors in `Range::kSDK`.
+    enum class SdkPart : uint8_t
+    {
+        kCore        = 0, ///< SDK core errors.
+        kInet        = 1, ///< Inet layer errors; see <inet/InetError.h>.
+        kDevice      = 2, ///< Device layer errors; see <platform/CHIPDeviceError.h>.
+        kASN1        = 3, ///< ASN1 errors; see <asn1/ASN1Error.h>.
+        kBLE         = 4, ///< BLE layer errors; see <ble/BleError.h>.
+        kApplication = 7, ///< Application-defined errors.
+    };
+
+    /// Test whether @a error belongs to @a range.
+    static constexpr bool IsRange(Range range, BaseType error)
+    {
+        return (error & MakeMask(kRangeStart, kRangeLength)) == MakeField(kRangeStart, static_cast<BaseType>(range));
+    }
+
+    static constexpr Range GetRange(BaseType error) { return static_cast<Range>(GetField(kRangeStart, kRangeLength, error)); }
+    static BaseType GetValue(BaseType error) { return GetField(kValueStart, kValueLength, error); }
+
+    /// Test whether if @a value can be losslessly encapsulated in a `CHIP_ERROR`.
+    static constexpr bool CanEncapsulate(BaseType value) { return FitsInField(kValueLength, value); }
+
+    /// Construct a `CHIP_ERROR` encapsulating @a value inside the @a range.
+    static BaseType Encapsulate(Range range, BaseType value)
+    {
+        return MakeInteger(range, (value & MakeMask(kValueStart, kValueLength)));
```

Comment:
So this is relying on kValueStart being 0, right?  Otherwise I'd expect to see:
```suggestion
        return MakeInteger(range, (value & MakeMask(0, kValueLength)));
```
Otherwise you are effectively left-shifting by kValueLength after taking the bits that are already shifted by kValueLength,